src/python/pants/backend/python/dependency_inference/rules.py

```diff
@@ -275,24 +310,84 @@ def _get_imports_info(
         )
         maybe_disambiguated = explicitly_provided_deps.disambiguated(owners.ambiguous)
         if maybe_disambiguated:
-            inferred_deps.add(maybe_disambiguated)
+            resolve_result[imp] = ImportResolveResult(
+                ImportOwnerStatus.disambiguated, (maybe_disambiguated,)
+            )
+        elif imp.split(".")[0] in DEFAULT_UNOWNED_DEPENDENCIES:
+            resolve_result[imp] = ImportResolveResult(ImportOwnerStatus.unownable)
+        elif parsed_imports[imp].weak:
+            resolve_result[imp] = ImportResolveResult(ImportOwnerStatus.weak_ignore)
+        else:
+            resolve_result[imp] = ImportResolveResult(ImportOwnerStatus.unowned)
+
+    return resolve_result
+
+
+def _collect_imports_info(
+    resolve_result: dict[str, ImportResolveResult]
+) -> tuple[frozenset[Address], frozenset[str]]:
+    """Collect import resolution results into:
+
+    - imports (direct and disambiguated)
+    - unowned
+    """
 
+    return frozenset(
+        addr
+        for dep in resolve_result.values()
+        for addr in dep.address
         if (
-            not owners.unambiguous
-            and imp.split(".")[0] not in DEFAULT_UNOWNED_DEPENDENCIES
-            and not parsed_imports[imp].weak
-        ):
-            unowned_imports.add(imp)
+            dep.status == ImportOwnerStatus.unambiguous
+            or dep.status == ImportOwnerStatus.disambiguated
+        )
+        and dep.address
+    ), frozenset(
+        imp for imp, dep in resolve_result.items() if dep.status == ImportOwnerStatus.unowned
+    )
+
+
+@dataclass(frozen=True)
+class UnownedImportsPossibleOwnersRequest:
+    """A request to find possible owners for several imports originating in a resolve."""
 
-    return inferred_deps, unowned_imports
+    unowned_imports: frozenset[str]
+    original_resolve: str
+
+
+@dataclass(frozen=True)
+class UnownedImportsPossibleOwners:
+    value: Dict[str, list[tuple[Address, ResolveName]]]
+
+
+@rule
+async def find_other_owners_for_unowned_imports(
+    req: UnownedImportsPossibleOwnersRequest,
+    python_setup: PythonSetup,
+) -> UnownedImportsPossibleOwners:
```

Comment:
Should this be a `@rule`, rather than a `@_rule_helper`? Is it likely to be called multiple times with the same inputs (such that it gains benefit from memoization)? If not, maybe `@_rule_helper`.

Reply:
hmm, I think that memoising individual possible owners would be useful (since it might be imported a few times), but not as much the whole bundle. I'll try rebuilding it.